pkg/helm/factory.go

```diff
@@ -0,0 +1,126 @@
+	Delete(ctx context.Context, kubeconfigFilePath, installName, namespace string) error
+	UpgradeChartWithValuesFile(ctx context.Context, chart, ociURI, version, kubeconfigFilePath, valuesFilePath string, opts ...Opt) error
+	InstallChartWithValuesFile(ctx context.Context, chart, ociURI, version, kubeconfigFilePath, valuesFilePath string) error
+	InstallChart(ctx context.Context, chart, ociURI, version, kubeconfigFilePath, namespace, valueFilePath string, skipCRDs bool, values []string) error
+	Template(ctx context.Context, ociURI, version, namespace string, values interface{}, kubeVersion string) ([]byte, error)
+	RegistryLogin(ctx context.Context, registry, username, password string) error
+}
+
+// WithRegistryMirror sets up registry mirror for helm.
+func WithRegistryMirror(mirror *registrymirror.RegistryMirror) Opt {
+	return func(c *Config) {
+		c.RegistryMirror = mirror
+	}
+}
+
+// WithInsecure configures helm to skip validating TLS certificates when
+// communicating with the Kubernetes API.
+func WithInsecure() Opt {
+	return func(c *Config) {
+		c.Insecure = true
+	}
+}
+
+// WithProxyConfig sets the proxy configurations on the Config.
+func WithProxyConfig(proxyConfig map[string]string) Opt {
+	return func(c *Config) {
+		c.ProxyConfig = proxyConfig
+	}
+}
+
+// ExecutableBuilder builds the Helm executable and returns it.
+type ExecutableBuilder interface {
+	BuildHelmExecutable(...Opt) Client
+}
+
+// ClientFactory provides a helm client for a cluster.
+type ClientFactory struct {
+	client         client.Client
+	helmClient     Client
+	mu             sync.Mutex
+	builder        ExecutableBuilder
+	registryMirror *registrymirror.RegistryMirror
+}
+
+// NewClientForClusterFactory returns a new helm ClientFactory.
+func NewClientForClusterFactory(client client.Client, builder ExecutableBuilder) *ClientFactory {
+	hf := &ClientFactory{
+		client:  client,
+		builder: builder,
+		mu:      sync.Mutex{},
+	}
+	return hf
+}
+
+// Get returns a new Helm client configured using information from the provided cluster's management cluster.
+func (f *ClientFactory) Get(ctx context.Context, clus *anywherev1.Cluster) (Client, error) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	managmentCluster := clus
+
+	var err error
+	if clus.IsManaged() {
+		managmentCluster, err = clusters.FetchManagementEksaCluster(ctx, f.client, clus)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var rUsername, rPassword string
+	if managmentCluster.RegistryAuth() {
+		rUsername, rPassword, err = configcli.ReadCredentialsFromSecret(ctx, f.client)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	r := registrymirror.FromCluster(managmentCluster)
+
+	// There is no need to rebuild the client everytime,
+	// here we check if the registry has changed and rebuild the helm client in that case
+	// Rebuild the Helm client only if the registry has changed
+	if f.helmClient == nil || !reflect.DeepEqual(f.registryMirror, r) {
```

Comment:
i advise against using deepequal in actual code, for tests is ok
is it difficult to write an `Equal` method for `registryMirror`?

Reply:
It's not difficult to write an equal method for `registryMirror`.

However, this is related to the thread [here](https://github.com/aws/eks-anywhere/pull/7170#discussion_r1428182609), so it may be removed

Reply:
We decided to remove this along with the client caching and locking mechanism.